### ISSUE TITLE
Is it safe to run multiple instances of pdfium-cli?

### ISSUE THREAD

**Issue by msageryd** [klippa-app/pdfium-cli#26]
I have stress tested pdfium-cli and started a bunch of Node child processes in parallel. It seems to work great, with no issues.
When I did the same with pdftocairo I sometimes got into problem where I got one or more corrupt jpg files as output.

Can you confirm that pdfium is safe to run in parallel?
Should I perform any specific tests to verify this?

**Comment by jerbob92** [comment]
Yes it is safe to run in parallel, I never had any issues with this. I have no idea why pdftocairo would have issues with that though. Multiple processes should have no interaction with each other. Only thing I can image causing issues is if you write to the same file at the same time from different processes.

**Comment by msageryd** [comment]
Thanks!

Regarding my previous timing test on different vCPU setups. Giving pdfium more than 2 vCPUs barely increased the performance, as you might recall. 

I haven't done any detailed test, but while stress testing pdfium it was obvious that multiple child processes helped with saturating more CPU cores. I would even guess that I can speed up some large jobs by dividing the pages string into multiple sets and give them to separate pdfium processes.

It seems fine for multiple pdfium processes to read from the same file. Do you know if this could give problems with the seek function? I suppose that seeking is handled completely from inside of pdfium, which would make multiple parallel reads safe as well, just asking to be sure.

**Comment by jerbob92** [comment]
> I haven't done any detailed test, but while stress testing pdfium it was obvious that multiple child processes helped with saturating more CPU cores. I would even guess that I can speed up some large jobs by dividing the pages string into multiple sets and give them to separate pdfium processes.

Since pdfium is not multi-threaded, the only way to saturate multiple cores in the rendering process you have to split up the jobs into smaller jobs indeed, there are some downsides to this:

- More memory usage, pdfium and (parts of) file live multiple times in RAM at the same time, if your PDF file will use a lot of memory, this might be problematic
- Time spent on loading, it has to open the same file multiple times so you will lose extra time starting the process, loading pdfium and opening the file

> It seems fine for multiple pdfium processes to read from the same file. Do you know if this could give problems with the seek function? I suppose that seeking is handled completely from inside of pdfium, which would make multiple parallel reads safe as well, just asking to be sure.

File seeking is handled by the OS/Filesystem and works on a [file descriptor](https://en.wikipedia.org/wiki/File_descriptor) base, so if you open the same file twice, the seeking will work fine in both processes, the file handles won't interfere with each other.

**Comment by msageryd** [comment]
More memory is cheaper than more containers on Fargate. I will not split up my jobs for now, but it's great to have the possibility. 

Most important is that I don't need to queue all jobs in a single line. If a new client request comes in while serving another client, I don't have to put the new client on hold. I can tell my worker how many max concurrent tasks to process. I'll start out with 3 for now.

Thanks again.
